Queries that fail due to bad connections are not automatically retried (problem with Issue #275 changes)

**Describe the bug**
Go's connection pool includes logic to discard bad connections and retry queries that failed due to a bad connection. This logic does not work with the mssql driver, because mssql does not return the standard driver.ErrBadConn error.

**To Reproduce**
Unfortunately, this bug cannot be reproduced with a simple code snippet, because it involves simulating transient network errors. A general guide to reproducing it is:

1. Start a debug session
2. Set a breakpoint on checkBadConn in mssql.go
3. Establish a database connection and execute a query via sql.go's DB.QueryContext
4. Observe that the query completes successfully
5. Disable your network
6. Repeat the query
7. Step through, but not out of, checkBadConn and observe that it returns a detailed error rather than the standard driver.ErrBadConn error
8. Re-enable your network
8. Step out of checkBadConn up to DB.QueryContext
9. Observe that the logic to discard the bad connection and retry with another is bypassed. The error propagates immediately to the caller

**Expected behavior**
1. mssql immediately returns driver.ErrBadConn when it encounters a bad connection, not later when the connection is used again
2. Go's connection pool logic is then able to immediately discard the bad connection and retry with a new connection

### Further technical details
SQL Server version: any
Operating system: any
Table schema: any

**Additional context**
I read through [Issue #275](https://github.com/denisenkom/go-mssqldb/issues/275) and understand the desire for detailed error messages. I also understand the limitations of Go's connection pooling and am disappointed that Go's [Issue #20807](https://github.com/golang/go/issues/20807) languished unresolved after hopeful initial attention.

However, the changes made to address #275 violate the contract with Go's connection pool, break Go's widely advertised automatic retry logic, and interfere with the pool's connection management (even though bad connections will eventually be purged). I believe that these costs are not worth the benefit of the detailed message.

Are the mssql maintainers amenable to rethinking Issue #275 and discussions of alternate approaches?

For handy reference, here is the code for sql.go's [QueryContext](https://github.com/golang/go/blob/dd150176c3cc49da68c8179f740eadc79404d351/src/database/sql/sql.go#L1592). You can see that by _not_ returning driver.ErrBadConn, mssql causes the connection pool logic to break out of the loop and return an error without making any retry attempts:

<pre>
func (db *DB) QueryContext(ctx context.Context, query string, args ...interface{}) (*Rows, error) {
	var rows *Rows
	var err error
	for i := 0; i < maxBadConnRetries; i++ {
		rows, err = db.query(ctx, query, args, cachedOrNewConn)
		<b>if err != driver.ErrBadConn {
			break</b>
		}
	}
	if err == driver.ErrBadConn {
		return db.query(ctx, query, args, alwaysNewConn)
	}
	return rows, err
</pre>

Similar retry logic exists throughout sql.go, and it is all broken by returning specific errors rather than driver.ErrBadConn.

Here is a unit test that demonstrates the problem using a single connection. It is modeled after the existing TestDisconnect1 and TestDisconnect2 tests in queries_test.go:

```
// TestDisconnect3 tests that a bad connection is removed from the pool and
// replaced with a new connection when a subsequent query executes
func TestDisconnect3(t *testing.T) {
	if testing.Short() {
		t.Skip("short")
	}
	checkConnStr(t)
	SetLogger(testLogger{t})

	// Revert to the normal dialer after the test is done.
	normalCreateDialer := createDialer
	defer func() {
		createDialer = normalCreateDialer
	}()

	var di *dialerInterrupt
	createDialer = func(p *connectParams) Dialer {
		nd := netDialer{&net.Dialer{Timeout: p.dial_timeout, KeepAlive: p.keepAlive}}
		di = &dialerInterrupt{nd: nd}
		return di
	}

	// Open a connection pool that holds a single connection to make sure all
	// queries run on the same connection.
	db, err := sql.Open("sqlserver", makeConnStr(t).String())
	if err != nil {
		t.Log(err)
	}
	defer db.Close()
	db.SetMaxIdleConns(1)
	db.SetMaxOpenConns(1)

	// Demonstrate connection is good the first time we use it
	var resultBefore string
	row := db.QueryRow("select 'before disconnect'")
	if err = row.Scan(&resultBefore); err != nil || resultBefore != "before disconnect" {
		t.Fatalf("Got result = '%s', err = %v, wanted result = 'before disconnect, err = <nil>", resultBefore, err)
	}

	// Break the connection
	di.Interrupt(true)
	di.Interrupt(false)

	// Broken connection should cause the next query to initially fail, but then
	// database/sql should discard the bad connection, automatically retry the query,
	// and ultimately succeed.
	var resultAfter string
	row = db.QueryRow("select 'after disconnect'")
	if err = row.Scan(&resultAfter); err != nil || resultAfter != "after disconnect" {
		t.Fatalf("Got result = '%s', err = %v, wanted result = 'after disconnect, err = <nil>", resultAfter, err)
	}
}
```

Here are the results of running it as go-msqldb exists today:

```
Running tool: /usr/local/go/bin/go test -timeout 30s -run ^TestDisconnect3$ github.com/denisenkom/go-mssqldb

--- FAIL: TestDisconnect3 (0.02s)
    /home/cfitzpatrick/go/src/go-mssqldb/tds_test.go:206: got token tokenLoginAck
    /home/cfitzpatrick/go/src/go-mssqldb/tds_test.go:206: got token tokenEnvChange
    /home/cfitzpatrick/go/src/go-mssqldb/tds_test.go:206: got token tokenDone
    /home/cfitzpatrick/go/src/go-mssqldb/tds_test.go:206: got DONE or DONEPROC status=0
    /home/cfitzpatrick/go/src/go-mssqldb/tds_test.go:210: select 'before disconnect'
    /home/cfitzpatrick/go/src/go-mssqldb/tds_test.go:206: got token tokenColMetadata
    /home/cfitzpatrick/go/src/go-mssqldb/tds_test.go:206: got token tokenRow
    /home/cfitzpatrick/go/src/go-mssqldb/tds_test.go:206: got token tokenDone
    /home/cfitzpatrick/go/src/go-mssqldb/tds_test.go:206: got DONE or DONEPROC status=16
    /home/cfitzpatrick/go/src/go-mssqldb/tds_test.go:206: (1 row(s) affected)
    /home/cfitzpatrick/go/src/go-mssqldb/tds_test.go:210: select 'after disconnect'
    /home/cfitzpatrick/go/src/go-mssqldb/tds_test.go:206: Failed to send SqlBatch with disconnect
    /home/cfitzpatrick/go/src/go-mssqldb/queries_test.go:2283: Got result = '', err = failed to send SQL Batch: disconnect, wanted result = 'after disconnect, err = <nil>
FAIL
FAIL	github.com/denisenkom/go-mssqldb	0.030s
FAIL
```

Here are the results running it with a version of go-msqldb modified to accept and use a BadConnAction=Retry connection string parameter (along the lines of kardianos' point three):

```
Running tool: /usr/local/go/bin/go test -timeout 30s -run ^TestDisconnect3$ github.com/denisenkom/go-mssqldb

ok  	github.com/denisenkom/go-mssqldb	0.065s
```

Here is an alternative unit test that uses tc-hib's idea of raising a fatal error in SQL Server. This may be a bit simpler to understand for someone who isn't familiar with the internal dialerInterrupt used in go-mssqldb unit tests:

```
// TestDisconnect4 tests that a bad connection is removed from the pool and
// replaced with a new connection when a subsequent query executes
func TestDisconnect4(t *testing.T) {
	if testing.Short() {
		t.Skip("short")
	}
	checkConnStr(t)
	SetLogger(testLogger{t})

	// Open a connection pool that holds a single connection to make sure all
	// queries run on the same connection.
	db, err := sql.Open("sqlserver", makeConnStr(t).String())
	if err != nil {
		t.Log(err)
	}
	defer db.Close()
	db.SetMaxIdleConns(1)
	db.SetMaxOpenConns(1)

	// Demonstrate connection is good the first time we use it
	var resultBefore string
	row := db.QueryRow("select 'before disconnect'")
	if err = row.Scan(&resultBefore); err != nil || resultBefore != "before disconnect" {
		t.Fatalf("Got result = '%s', err = %v, wanted result = 'before disconnect, err = <nil>", resultBefore, err)
	}

	// Break the connection by raising a fatal error in SQL Server. When you raise an
	// error of severity 20 or higher, SQL Server closes the client connection.
	res, err := db.Exec("raiserror('goodbye', 20, 1) with log")
	if err == nil || err.Error() != "SQL Server had internal error" {
		t.Fatalf("Failed to break connection to SQL database. Got err = %v, wanted error = 'SQL Server had internal error'", err)
	}
	t.Log(res)

	// Broken connection should cause the next query to initially fail, but then
	// database/sql should discard the bad connection, automatically retry the query,
	// and ultimately succeed.
	var resultAfter string
	row = db.QueryRow("select 'after disconnect'")
	if err = row.Scan(&resultAfter); err != nil || resultAfter != "after disconnect" {
		t.Fatalf("Got result = '%s', err = %v, wanted result = 'after disconnect, err = <nil>", resultAfter, err)
	}
}
```

Results with current version of go-mssqldb:

```
Running tool: /usr/local/go/bin/go test -timeout 30s -run ^TestDisconnect4$ github.com/denisenkom/go-mssqldb

--- FAIL: TestDisconnect4 (0.02s)
    /home/ctest/go/src/go-mssqldb/tds_test.go:206: got token tokenLoginAck
    /home/ctest/go/src/go-mssqldb/tds_test.go:206: got token tokenEnvChange
    /home/ctest/go/src/go-mssqldb/tds_test.go:206: got token tokenDone
    /home/ctest/go/src/go-mssqldb/tds_test.go:206: got DONE or DONEPROC status=0
    /home/ctest/go/src/go-mssqldb/tds_test.go:210: select 'before disconnect'
    /home/ctest/go/src/go-mssqldb/tds_test.go:206: got token tokenColMetadata
    /home/ctest/go/src/go-mssqldb/tds_test.go:206: got token tokenRow
    /home/ctest/go/src/go-mssqldb/tds_test.go:206: got token tokenDone
    /home/ctest/go/src/go-mssqldb/tds_test.go:206: got DONE or DONEPROC status=16
    /home/ctest/go/src/go-mssqldb/tds_test.go:206: (1 row(s) affected)
    /home/ctest/go/src/go-mssqldb/tds_test.go:210: raiserror('goodbye', 20, 1) with log
    /home/ctest/go/src/go-mssqldb/tds_test.go:206: got token tokenEnvChange
    /home/ctest/go/src/go-mssqldb/tds_test.go:206: got token tokenError
    /home/ctest/go/src/go-mssqldb/tds_test.go:206: got ERROR 50000 goodbye
    /home/ctest/go/src/go-mssqldb/tds_test.go:210: goodbye
    /home/ctest/go/src/go-mssqldb/tds_test.go:206: got token tokenError
    /home/ctest/go/src/go-mssqldb/tds_test.go:206: got ERROR 2745 Process ID 55 has raised user error 50000, severity 20. SQL Server is terminating this process.
    /home/ctest/go/src/go-mssqldb/tds_test.go:210: Process ID 55 has raised user error 50000, severity 20. SQL Server is terminating this process.
    /home/ctest/go/src/go-mssqldb/tds_test.go:206: got token tokenError
    /home/ctest/go/src/go-mssqldb/tds_test.go:206: got ERROR 596 Cannot continue the execution because the session is in the kill state.
    /home/ctest/go/src/go-mssqldb/tds_test.go:210: Cannot continue the execution because the session is in the kill state.
    /home/ctest/go/src/go-mssqldb/tds_test.go:206: got token tokenDone
    /home/ctest/go/src/go-mssqldb/tds_test.go:206: got DONE or DONEPROC status=258
    /home/ctest/go/src/go-mssqldb/queries_test.go:2318: <nil>
    /home/ctest/go/src/go-mssqldb/tds_test.go:210: select 'after disconnect'
    /home/ctest/go/src/go-mssqldb/tds_test.go:206: Failed to send SqlBatch with write tcp 192.168.32.130:60214->192.168.32.129:1433: write: connection reset by peer
    /home/ctest/go/src/go-mssqldb/queries_test.go:2326: Got result = '', err = failed to send SQL Batch: write tcp 192.168.32.130:60214->192.168.32.129:1433: write: connection reset by peer, wanted result = 'after disconnect, err = <nil>
FAIL
FAIL	github.com/denisenkom/go-mssqldb	0.030s
FAIL
```

Results with modified version that auto-retries:

```
Running tool: /usr/local/go/bin/go test -timeout 30s -run ^TestDisconnect4$ github.com/denisenkom/go-mssqldb

ok  	github.com/denisenkom/go-mssqldb	0.060s
```

I added unit tests and a DSN parameter that allows configuration of auto-retry behavior on top of tc-hib's commit 405258fae9b25739c319145fe9f51c9ac6b78b67. You can see my changes on my fork at commit 5ad7cee471dfe955b04ecdd9de49c3af7b84d763.

Tc-hib's changes failed the existing unit tests TestShortTimeout and TestQueryTimeout because the changes wrapped the original errors in msql.Error errors:

```
--- FAIL: TestShortTimeout (2.02s)
    ...
    queries_test.go:566: failure not a timeout, failed with mssql: read tcp 192.168.32.130:48322->192.168.32.129:1433: i/o timeout
--- FAIL: TestQueryTimeout (0.24s)
    ...
    queries_test.go:1965: ExecContext expected to fail with DeadlineExceeded but it returned mssql: context deadline exceeded
```

I think, based on the fields and functions of mssql.Error and the fact that there is also an existing StreamError struct, that Error was meant exclusively for errors returned by SQL Server. Therefore, we don't want to overload it and return it for all errors when a connection is bad.

Instead, I propose creating a new RetryableError for the exclusive purpose of conveying information about errors that can be retried. RetryableError implements errors.Is, returning true when compared to driver.ErrBadConn to enable database/sql's retry logic. It also wraps the original error so that the caller can get the root cause if the automatic retry logic fails. RetryableError is only used within checkBadConn, where I believe it simplifies and clarifies the logic. It also obviates the need to include a badConn field in mssql.Error, which may be out of place there, because SQL Server errors don't generally have anything to do with connection state.

I think we are getting pretty close to where we want to be, but we still need to:

1. Decide what the default retry logic should be going forward:. Do we want it to be "no retry" in the style desired in #275? Or do we want it to be "auto retry" in the style desired in #586? Keep in mind that "auto retry", although the standard and expected behavior within database/sql, is not the current behavior of go-mssqldb.
2. Create a version that works with or without kardianos' proposed change to database/sql (presumably by creating two versions of checkBadConn in separate files gated by build tags/constraints).

What are everyone's thoughts?

Also, is it appropriate to carry on this level of discussion in the issue comments? If not, is GitHub's discussion feature the appropriate venue (in which case someone would have to enable it in this repository)?

--- a/bulkcopy.go
+++ b/bulkcopy.go
@@ -238,7 +238,7 @@ func (b *Bulk) Done() (rowcount int64, err error) {
 	reader := startReading(b.cn.sess, b.ctx, nil)
 	err = reader.iterateResponse()
 	if err != nil {
-		return 0, b.cn.checkBadConn(err)
+		return 0, b.cn.checkBadConn(err, false)
 	}
 
 	return reader.rowCount, nil

--- a/error.go
+++ b/error.go
@@ -1,6 +1,7 @@
 package mssql
 
 import (
+	"database/sql/driver"
 	"fmt"
 )
 
@@ -20,6 +21,17 @@ type Error struct {
 	// All lists all errors that were received from first to last.
 	// This includes the last one, which is described in the other members.
 	All []Error
+	// badConn means the error allows the connection pool
+	// to drop the connection and immediately retry
+	// querying on another connection.
+	badConn bool
+}
+
+func (e Error) Unwrap() error {
+	if e.badConn {
+		return driver.ErrBadConn
+	}
+	return nil
 }
 
 func (e Error) Error() string {

--- a/error.go
+++ b/error.go
@@ -21,17 +21,6 @@ type Error struct {
 	// All lists all errors that were received from first to last.
 	// This includes the last one, which is described in the other members.
 	All []Error
-	// badConn means the error allows the connection pool
-	// to drop the connection and immediately retry
-	// querying on another connection.
-	badConn bool
-}
-
-func (e Error) Unwrap() error {
-	if e.badConn {
-		return driver.ErrBadConn
-	}
-	return nil
 }
 
 func (e Error) Error() string {
@@ -87,10 +76,12 @@ func badStreamPanicf(format string, v ...interface{}) {
 	panic(streamErrorf(format, v...))
 }
 
-// ServerError is returned when the server got a fatal error.
+// ServerError is returned when the server got a fatal error
+// that aborts the process and severs the connection.
 //
 // To get the errors returned before the process was aborted,
-// use errors.As with an *mssql.Error.
+// unwrap this error or call errors.As with a pointer to an
+// mssql.Error variable.
 type ServerError struct {
 	sqlError Error
 }
@@ -101,4 +92,30 @@ func (e ServerError) Error() string {
 
 func (e ServerError) Unwrap() error {
 	return e.sqlError
+}
+
+// RetryableError is returned when an error was caused by a bad
+// connection at the start of a query and can be safely retried
+// using database/sql's automatic retry logic.
+//
+// In many cases database/sql's retry logic will transparently
+// handle this error, the retried call will return successfully,
+// and you won't even see this error. However, you may see this
+// error if the retry logic cannot successfully handle the error.
+// In that case you can get the underlying error by calling this
+// error's UnWrap function.
+type RetryableError struct {
+	err error
+}
+
+func (r RetryableError) Error() string {
+	return r.err.Error()
+}
+
+func (r RetryableError) Unwrap() error {
+	return r.err
+}
+
+func (r RetryableError) Is(err error) bool {
+	return err == driver.ErrBadConn
 }

--- a/error.go
+++ b/error.go
@@ -69,7 +69,7 @@ func streamErrorf(format string, v ...interface{}) StreamError {
 }
 
 func badStreamPanic(err error) {
-	panic(err)
+	panic(streamErrorf("%v", err))
 }
 
 func badStreamPanicf(format string, v ...interface{}) {

--- a/error.go
+++ b/error.go
@@ -74,3 +74,19 @@ func badStreamPanic(err error) {
 func badStreamPanicf(format string, v ...interface{}) {
 	panic(streamErrorf(format, v...))
 }
+
+// ServerError is returned when the server got a fatal error.
+//
+// To get the errors returned before the process was aborted,
+// use errors.As with an *mssql.Error.
+type ServerError struct {
+	sqlError Error
+}
+
+func (e ServerError) Error() string {
+	return "SQL Server had internal error"
+}
+
+func (e ServerError) Unwrap() error {
+	return e.sqlError
+}

--- a/error_test.go
+++ b/error_test.go
@@ -1,0 +1,87 @@
+package mssql
+
+import (
+	"database/sql/driver"
+	"fmt"
+	"strings"
+	"testing"
+)
+
+func TestServerError(t *testing.T) {
+
+	originalErr := Error{Message: "underlying error"}
+	sererErr := ServerError{sqlError: originalErr}
+
+	// Verify that error message is backwards compatible
+	oldMessage := "SQL Server had internal error"
+	if newMessage := sererErr.Error(); newMessage != oldMessage {
+		t.Fatalf("ServerError returned incompatible error message. Got '%s', wanted '%s'", newMessage, oldMessage)
+	}
+
+	// Verify that the underlying error is preserved
+	unwrappedErr := sererErr.Unwrap()
+	if underlyingErr, ok := unwrappedErr.(Error); !ok || underlyingErr.Message != originalErr.Message {
+		t.Fatalf("ServerError did not preserve wrapped error. Got '%+v', wanted '%+v'", unwrappedErr, originalErr)
+	}
+}
+
+func TestRetryableError(t *testing.T) {
+
+	originalErr := driver.ErrBadConn
+	retryableErr := RetryableError{err: originalErr}
+
+	// Verify that the error message matches the original error's
+	origMessage := originalErr.Error()
+	if wrappedMessage := retryableErr.Error(); wrappedMessage != origMessage {
+		t.Fatalf("RetryableError returned incorrect error message. Got '%s', wanted '%s'", wrappedMessage, origMessage)
+	}
+
+	// Verify that the underlying error is preserved
+	unwrappedErr := retryableErr.Unwrap()
+	if unwrappedErr != originalErr {
+		t.Fatalf("RetryableError did not preserve wrapped error. Got '%+v', wanted '%+v'", unwrappedErr, originalErr)
+	}
+
+	// Verify that underlying error is correctly recognized
+	if !retryableErr.Is(driver.ErrBadConn) {
+		t.Fatalf("RetryableError wrapping driver.ErrBadConn does not report it is a driver.ErrBadConn error")
+	}
+
+}
+
+func TestBadStreamPanic(t *testing.T) {
+
+	errMsg := "test error XYZ"
+	err := fmt.Errorf(errMsg)
+
+	defer func() {
+		r := recover()
+		if e, ok := r.(error); !ok || !strings.HasSuffix(e.Error(), errMsg) {
+			t.Fatalf("unexpected error recovered from panic: "+
+				"got error = '%+v', wanted error to end with '%s'", e, errMsg)
+		}
+	}()
+
+	badStreamPanic(err)
+
+	t.Fatalf("badStreamPanic did not panic as expected when passed %+v", err)
+}
+
+func TestBadStreamPanicf(t *testing.T) {
+
+	errfmt := "the error is '%s'"
+	errMsg := "test error XYZ"
+	expectedMsg := fmt.Sprintf(errfmt, errMsg)
+
+	defer func() {
+		r := recover()
+		if e, ok := r.(error); !ok || !strings.HasSuffix(e.Error(), expectedMsg) {
+			t.Fatalf("unexpected error recovered from panic: "+
+				"got error = '%+v', wanted error to end with '%s'", e, expectedMsg)
+		}
+	}()
+
+	badStreamPanicf(errfmt, errMsg)
+
+	t.Fatalf("badStreamPanicf did not panic as expected when passed %s", expectedMsg)
+}

--- a/msdsn/conn_str.go
+++ b/msdsn/conn_str.go
@@ -61,6 +61,10 @@ type Config struct {
 	Workstation string
 	AppName     string
 
+	// If true disables database/sql's automatic retry of queries
+	// that start on bad connections.
+	DisableRetry bool
+
 	// Do not use the following.
 
 	DialTimeout time.Duration // DialTimeout defaults to 15s. Set negative to disable.
@@ -293,8 +297,18 @@ func Parse(dsn string) (Config, map[string]string, error) {
 		var err error
 		p.FailOverPort, err = strconv.ParseUint(failOverPort, 0, 16)
 		if err != nil {
-			f := "invalid tcp port '%v': %v"
+			f := "invalid failover port '%v': %v"
 			return p, params, fmt.Errorf(f, failOverPort, err.Error())
+		}
+	}
+
+	disableRetry, ok := params["disableretry"]
+	if ok {
+		var err error
+		p.DisableRetry, err = strconv.ParseBool(disableRetry)
+		if err != nil {
+			f := "invalid disableRetry '%s': %s"
+			return p, params, fmt.Errorf(f, disableRetry, err.Error())
 		}
 	}
 
@@ -314,6 +328,9 @@ func (p Config) URL() *url.URL {
 	host := p.Host
 	if p.Port > 0 {
 		host = fmt.Sprintf("%s:%d", p.Host, p.Port)
+	}
+	if p.DisableRetry {
+		q.Add("disableRetry", "true")
 	}
 	res := url.URL{
 		Scheme: "sqlserver",

--- a/msdsn/conn_str.go
+++ b/msdsn/conn_str.go
@@ -310,6 +310,8 @@ func Parse(dsn string) (Config, map[string]string, error) {
 			f := "invalid disableRetry '%s': %s"
 			return p, params, fmt.Errorf(f, disableRetry, err.Error())
 		}
+	} else {
+		p.DisableRetry = disableRetryDefault
 	}
 
 	return p, params, nil
@@ -329,9 +331,7 @@ func (p Config) URL() *url.URL {
 	if p.Port > 0 {
 		host = fmt.Sprintf("%s:%d", p.Host, p.Port)
 	}
-	if p.DisableRetry {
-		q.Add("disableRetry", "true")
-	}
+	q.Add("disableRetry", fmt.Sprintf("%t", p.DisableRetry))
 	res := url.URL{
 		Scheme: "sqlserver",
 		Host:   host,

--- a/msdsn/conn_str_go118.go
+++ b/msdsn/conn_str_go118.go
@@ -1,0 +1,9 @@
+// +build go1.18
+
+package msdsn
+
+// disableRetryDefault is false for Go versions 1.18 and higher. This matches
+// the behavior requested in issue #586. A query that fails at the start due to
+// a bad connection is automatically retried. An error is returned only if the
+// query fails all of its retries.
+const disableRetryDefault bool = false

--- a/msdsn/conn_str_go118pre.go
+++ b/msdsn/conn_str_go118pre.go
@@ -1,0 +1,9 @@
+// +build !go1.18
+
+package msdsn
+
+// disableRetryDefault is true for versions of Go less than 1.18. This matches
+// the behavior requested in issue #275. A query that fails at the start due to
+// a bad connection is not retried. Instead, the detailed error is immediately
+// returned to the caller.
+const disableRetryDefault bool = true

--- a/msdsn/conn_str_test.go
+++ b/msdsn/conn_str_test.go
@@ -75,6 +75,7 @@ func TestValidConnectionString(t *testing.T) {
 		{"disableretry=false", func(p Config) bool { return !p.DisableRetry }},
 		{"disableretry=1", func(p Config) bool { return p.DisableRetry }},
 		{"disableretry=0", func(p Config) bool { return !p.DisableRetry }},
+		{"", func(p Config) bool { return p.DisableRetry == disableRetryDefault }},
 
 		// those are supported currently, but maybe should not be
 		{"someparam", func(p Config) bool { return true }},

--- a/msdsn/conn_str_test.go
+++ b/msdsn/conn_str_test.go
@@ -18,6 +18,7 @@ func TestInvalidConnectionString(t *testing.T) {
 		"trustservercertificate=invalid",
 		"failoverport=invalid",
 		"applicationintent=ReadOnly",
+		"disableretry=invalid",
 
 		// ODBC mode
 		"odbc:password={",
@@ -70,6 +71,10 @@ func TestValidConnectionString(t *testing.T) {
 		{"log=64;packet size=300", func(p Config) bool { return p.LogFlags == 64 && p.PacketSize == 512 }},
 		{"log=64;packet size=8192", func(p Config) bool { return p.LogFlags == 64 && p.PacketSize == 8192 }},
 		{"log=64;packet size=48000", func(p Config) bool { return p.LogFlags == 64 && p.PacketSize == 32767 }},
+		{"disableretry=true", func(p Config) bool { return p.DisableRetry }},
+		{"disableretry=false", func(p Config) bool { return !p.DisableRetry }},
+		{"disableretry=1", func(p Config) bool { return p.DisableRetry }},
+		{"disableretry=0", func(p Config) bool { return !p.DisableRetry }},
 
 		// those are supported currently, but maybe should not be
 		{"someparam", func(p Config) bool { return true }},
@@ -121,6 +126,12 @@ func TestValidConnectionString(t *testing.T) {
 		{"odbc:password={value}  ", func(p Config) bool {
 			return p.Password == "value"
 		}},
+		{"odbc:server=somehost;user id=someuser;password=somepass;disableretry=true", func(p Config) bool {
+			return p.Host == "somehost" && p.User == "someuser" && p.Password == "somepass" && p.DisableRetry
+		}},
+		{"odbc:server=somehost;user id=someuser;password=somepass; disableretry =  1 ", func(p Config) bool {
+			return p.Host == "somehost" && p.User == "someuser" && p.Password == "somepass" && p.DisableRetry
+		}},
 
 		// URL mode
 		{"sqlserver://somehost?connection+timeout=30", func(p Config) bool {
@@ -140,6 +151,12 @@ func TestValidConnectionString(t *testing.T) {
 		}},
 		{"sqlserver://someuser:foo%3A%2F%5C%21~%40;bar@somehost:1434/someinstance?connection+timeout=30", func(p Config) bool {
 			return p.Host == "somehost" && p.Port == 1434 && p.Instance == "someinstance" && p.User == "someuser" && p.Password == "foo:/\\!~@;bar" && p.ConnTimeout == 30*time.Second
+		}},
+		{"sqlserver://someuser@somehost?disableretry=true", func(p Config) bool {
+			return p.Host == "somehost" && p.Port == 0 && p.Instance == "" && p.User == "someuser" && p.Password == "" && p.DisableRetry
+		}},
+		{"sqlserver://someuser@somehost?connection+timeout=30&disableretry=1", func(p Config) bool {
+			return p.Host == "somehost" && p.Port == 0 && p.Instance == "" && p.User == "someuser" && p.Password == "" && p.ConnTimeout == 30*time.Second && p.DisableRetry
 		}},
 	}
 	for _, ts := range connStrings {

--- a/mssql.go
+++ b/mssql.go
@@ -190,10 +190,6 @@ func (c *Conn) checkBadConn(err error, mayRetry bool) error {
 		return nil
 	case io.EOF:
 		c.connectionGood = false
-		return Error{
-			Message: "unexpected EOF",
-			badConn: mayRetry,
-		}
 	case driver.ErrBadConn:
 		// It is an internal programming error if driver.ErrBadConn
 		// is ever passed to this function. driver.ErrBadConn should
@@ -202,22 +198,18 @@ func (c *Conn) checkBadConn(err error, mayRetry bool) error {
 		panic("driver.ErrBadConn in checkBadConn. This should not happen.")
 	}
 
-	switch err := err.(type) {
-	case ServerError:
+	switch err.(type) {
+	case net.Error:
 		c.connectionGood = false
-		err.sqlError.badConn = mayRetry
-		return err
-	case Error:
-		err.badConn = mayRetry && !c.connectionGood
-		return err
-	case net.Error, StreamError:
+	case StreamError:
+		c.connectionGood = false
+	case ServerError:
 		c.connectionGood = false
 	}
 
-	if !c.connectionGood {
-		return Error{
-			Message: err.Error(),
-			badConn: mayRetry,
+	if !c.connectionGood && mayRetry {
+		return RetryableError{
+			err: err,
 		}
 	}
 	return err

--- a/mssql.go
+++ b/mssql.go
@@ -214,11 +214,12 @@ func (c *Conn) checkBadConn(err error, mayRetry bool) error {
 			badConn: mayRetry,
 		}
 	case Error:
-		err.badConn = mayRetry
+		err.badConn = mayRetry && !c.connectionGood
 		return err
 	}
 
 	if mayRetry && !c.connectionGood {
+		fmt.Println("returning ErrBadConn!")
 		return driver.ErrBadConn
 	}
 	return err

--- a/mssql.go
+++ b/mssql.go
@@ -181,23 +181,18 @@ func (c *Conn) IsValid() bool {
 	return c.connectionGood
 }
 
-func (c *Conn) checkBadConn(err error) error {
-	// this is a hack to address Issue #275
-	// we set connectionGood flag to false if
-	// error indicates that connection is not usable
-	// but we return actual error instead of ErrBadConn
-	// this will cause connection to stay in a pool
-	// but next request to this connection will return ErrBadConn
-
-	// it might be possible to revise this hack after
-	// https://github.com/golang/go/issues/20807
-	// is implemented
+// checkBadConn checks if an error means the connection must be dropped.
+// If canRetry is true, it will return driver.ErrBadConn to allow the
+// the connection pool to retry with another connection.
+func (c *Conn) checkBadConn(err error, mayRetry bool) error {
 	switch err {
 	case nil:
 		return nil
+	case serverError:
+		c.connectionGood = false
 	case io.EOF:
 		c.connectionGood = false
-		return driver.ErrBadConn
+		err = driver.ErrBadConn
 	case driver.ErrBadConn:
 		// It is an internal programming error if driver.ErrBadConn
 		// is ever passed to this function. driver.ErrBadConn should
@@ -209,13 +204,14 @@ func (c *Conn) checkBadConn(err error) error {
 	switch err.(type) {
 	case net.Error:
 		c.connectionGood = false
-		return err
 	case StreamError:
 		c.connectionGood = false
-		return err
-	default:
-		return err
 	}
+
+	if mayRetry && !c.connectionGood {
+		return driver.ErrBadConn
+	}
+	return err
 }
 
 func (c *Conn) clearOuts() {
@@ -229,7 +225,7 @@ func (c *Conn) simpleProcessResp(ctx context.Context) error {
 	var resultError error
 	err := reader.iterateResponse()
 	if err != nil {
-		return c.checkBadConn(err)
+		return c.checkBadConn(err, false)
 	}
 	return resultError
 }
@@ -239,7 +235,7 @@ func (c *Conn) Commit() error {
 		return driver.ErrBadConn
 	}
 	if err := c.sendCommitRequest(); err != nil {
-		return c.checkBadConn(err)
+		return c.checkBadConn(err, true)
 	}
 	return c.simpleProcessResp(c.transactionCtx)
 }
@@ -266,7 +262,7 @@ func (c *Conn) Rollback() error {
 		return driver.ErrBadConn
 	}
 	if err := c.sendRollbackRequest(); err != nil {
-		return c.checkBadConn(err)
+		return c.checkBadConn(err, true)
 	}
 	return c.simpleProcessResp(c.transactionCtx)
 }
@@ -298,7 +294,7 @@ func (c *Conn) begin(ctx context.Context, tdsIsolation isoLevel) (tx driver.Tx, 
 	}
 	err = c.sendBeginRequest(ctx, tdsIsolation)
 	if err != nil {
-		return nil, c.checkBadConn(err)
+		return nil, c.checkBadConn(err, true)
 	}
 	tx, err = c.processBeginResponse(ctx)
 	if err != nil {
@@ -613,7 +609,7 @@ func (s *Stmt) queryContext(ctx context.Context, args []namedValue) (rows driver
 		return nil, driver.ErrBadConn
 	}
 	if err = s.sendQuery(args); err != nil {
-		return nil, s.c.checkBadConn(err)
+		return nil, s.c.checkBadConn(err, true)
 	}
 	return s.processQueryResponse(ctx)
 }
@@ -646,7 +642,7 @@ loop:
 					if token.isError() {
 						// need to cleanup cancellable context
 						cancel()
-						return nil, s.c.checkBadConn(token.getError())
+						return nil, s.c.checkBadConn(token.getError(), false)
 					}
 				case ReturnStatus:
 					s.c.sess.setReturnStatus(token)
@@ -655,7 +651,7 @@ loop:
 		} else {
 			// need to cleanup cancellable context
 			cancel()
-			return nil, s.c.checkBadConn(err)
+			return nil, s.c.checkBadConn(err, false)
 		}
 	}
 	res = &Rows{stmt: s, reader: reader, cols: cols, cancel: cancel}
@@ -671,7 +667,7 @@ func (s *Stmt) exec(ctx context.Context, args []namedValue) (res driver.Result, 
 		return nil, driver.ErrBadConn
 	}
 	if err = s.sendQuery(args); err != nil {
-		return nil, s.c.checkBadConn(err)
+		return nil, s.c.checkBadConn(err, true)
 	}
 	if res, err = s.processExec(ctx); err != nil {
 		return nil, err
@@ -684,7 +680,7 @@ func (s *Stmt) processExec(ctx context.Context) (res driver.Result, err error) {
 	s.c.clearOuts()
 	err = reader.iterateResponse()
 	if err != nil {
-		return nil, s.c.checkBadConn(err)
+		return nil, s.c.checkBadConn(err, false)
 	}
 	return &Result{s.c, reader.rowCount}, nil
 }
@@ -754,7 +750,7 @@ func (rc *Rows) Next(dest []driver.Value) error {
 					return nil
 				case doneStruct:
 					if tokdata.isError() {
-						return rc.stmt.c.checkBadConn(tokdata.getError())
+						return rc.stmt.c.checkBadConn(tokdata.getError(), false)
 					}
 				case ReturnStatus:
 					rc.stmt.c.sess.setReturnStatus(tokdata)
@@ -762,7 +758,7 @@ func (rc *Rows) Next(dest []driver.Value) error {
 			}
 
 		} else {
-			return rc.stmt.c.checkBadConn(err)
+			return rc.stmt.c.checkBadConn(err, false)
 		}
 	}
 }

--- a/mssql.go
+++ b/mssql.go
@@ -212,10 +212,9 @@ func (c *Conn) checkBadConn(err error, mayRetry bool) error {
 	}
 
 	if !c.connectionGood && mayRetry && !c.connector.params.DisableRetry {
-		return RetryableError{
-			err: err,
-		}
+		return newRetryableError(err)
 	}
+
 	return err
 }
 

--- a/mssql.go
+++ b/mssql.go
@@ -188,8 +188,6 @@ func (c *Conn) checkBadConn(err error, mayRetry bool) error {
 	switch err {
 	case nil:
 		return nil
-	case serverError:
-		c.connectionGood = false
 	case io.EOF:
 		c.connectionGood = false
 		err = driver.ErrBadConn
@@ -202,9 +200,7 @@ func (c *Conn) checkBadConn(err error, mayRetry bool) error {
 	}
 
 	switch err.(type) {
-	case net.Error:
-		c.connectionGood = false
-	case StreamError:
+	case serverError, net.Error, StreamError:
 		c.connectionGood = false
 	}
 

--- a/mssql.go
+++ b/mssql.go
@@ -209,18 +209,16 @@ func (c *Conn) checkBadConn(err error, mayRetry bool) error {
 		return err
 	case net.Error, StreamError:
 		c.connectionGood = false
-		return Error{
-			Message: err.Error(),
-			badConn: mayRetry,
-		}
 	case Error:
 		err.badConn = mayRetry && !c.connectionGood
 		return err
 	}
 
-	if mayRetry && !c.connectionGood {
-		fmt.Println("returning ErrBadConn!")
-		return driver.ErrBadConn
+	if !c.connectionGood {
+		return Error{
+			Message: err.Error(),
+			badConn: mayRetry,
+		}
 	}
 	return err
 }

--- a/mssql.go
+++ b/mssql.go
@@ -200,7 +200,7 @@ func (c *Conn) checkBadConn(err error, mayRetry bool) error {
 	}
 
 	switch err.(type) {
-	case serverError, net.Error, StreamError:
+	case ServerError, net.Error, StreamError:
 		c.connectionGood = false
 	}
 

--- a/mssql.go
+++ b/mssql.go
@@ -181,9 +181,13 @@ func (c *Conn) IsValid() bool {
 	return c.connectionGood
 }
 
-// checkBadConn checks if an error means the connection must be dropped.
-// If canRetry is true, it will return driver.ErrBadConn to allow the
-// the connection pool to retry with another connection.
+// checkBadConn marks the connection as bad based on the characteristics
+// of the suplied error. Bad connections will be dropped from the connection
+// pool rather than reused.
+//
+// If bad connection retry is enabled and the error + connection state permits
+// retrying, checkBadConn will return a RetryableError that allows database/sql
+// to automatically retry the query with another connection.
 func (c *Conn) checkBadConn(err error, mayRetry bool) error {
 	switch err {
 	case nil:
@@ -207,7 +211,7 @@ func (c *Conn) checkBadConn(err error, mayRetry bool) error {
 		c.connectionGood = false
 	}
 
-	if !c.connectionGood && mayRetry {
+	if !c.connectionGood && mayRetry && !c.connector.params.DisableRetry {
 		return RetryableError{
 			err: err,
 		}

--- a/mssql.go
+++ b/mssql.go
@@ -207,11 +207,11 @@ func (c *Conn) checkBadConn(err error, mayRetry bool) error {
 		c.connectionGood = false
 		err.sqlError.badConn = mayRetry
 		return err
-	case net.Error, StreamError:
-		c.connectionGood = false
 	case Error:
 		err.badConn = mayRetry && !c.connectionGood
 		return err
+	case net.Error, StreamError:
+		c.connectionGood = false
 	}
 
 	if !c.connectionGood {

--- a/mssql.go
+++ b/mssql.go
@@ -190,7 +190,10 @@ func (c *Conn) checkBadConn(err error, mayRetry bool) error {
 		return nil
 	case io.EOF:
 		c.connectionGood = false
-		err = driver.ErrBadConn
+		return Error{
+			Message: "unexpected EOF",
+			badConn: mayRetry,
+		}
 	case driver.ErrBadConn:
 		// It is an internal programming error if driver.ErrBadConn
 		// is ever passed to this function. driver.ErrBadConn should
@@ -199,9 +202,20 @@ func (c *Conn) checkBadConn(err error, mayRetry bool) error {
 		panic("driver.ErrBadConn in checkBadConn. This should not happen.")
 	}
 
-	switch err.(type) {
-	case ServerError, net.Error, StreamError:
+	switch err := err.(type) {
+	case ServerError:
 		c.connectionGood = false
+		err.sqlError.badConn = mayRetry
+		return err
+	case net.Error, StreamError:
+		c.connectionGood = false
+		return Error{
+			Message: err.Error(),
+			badConn: mayRetry,
+		}
+	case Error:
+		err.badConn = mayRetry
+		return err
 	}
 
 	if mayRetry && !c.connectionGood {

--- a/mssql.go
+++ b/mssql.go
@@ -182,7 +182,7 @@ func (c *Conn) IsValid() bool {
 }
 
 // checkBadConn marks the connection as bad based on the characteristics
-// of the suplied error. Bad connections will be dropped from the connection
+// of the supplied error. Bad connections will be dropped from the connection
 // pool rather than reused.
 //
 // If bad connection retry is enabled and the error + connection state permits

--- a/mssql_go118.go
+++ b/mssql_go118.go
@@ -1,0 +1,14 @@
+// +build go1.18
+
+package mssql
+
+// newRetryableError returns an error that allows the database/sql package
+// to automatically retry the failed query. Versions of Go 1.18 and higher
+// use errors.Is to determine whether or not a failed query can be retried.
+// Therefore, we wrap the underlying error in a RetryableError that both
+// implements errors.Is for automatic retry and maintains the error details.
+func newRetryableError(err error) error {
+	return RetryableError{
+		err: err,
+	}
+}

--- a/mssql_go118pre.go
+++ b/mssql_go118pre.go
@@ -1,0 +1,17 @@
+// +build !go1.18
+
+package mssql
+
+import (
+	"database/sql/driver"
+)
+
+// newRetryableError returns an error that allows the database/sql package
+// to automatically retry the failed query. Versions of Go lower than 1.18
+// compare directly to the sentinel error driver.ErrBadConn to determine
+// whether or not a failed query can be retried. Therefore, we replace the
+// actual error with driver.ErrBadConn, enabling retry but losing the error
+// details.
+func newRetryableError(err error) error {
+	return driver.ErrBadConn
+}

--- a/mssql_test.go
+++ b/mssql_test.go
@@ -3,7 +3,14 @@ package mssql
 import (
 	"context"
 	"database/sql"
+	"database/sql/driver"
+	"fmt"
+	"io"
+	"net"
+	"reflect"
 	"testing"
+
+	"github.com/denisenkom/go-mssqldb/msdsn"
 )
 
 func TestBadOpen(t *testing.T) {
@@ -68,4 +75,81 @@ func TestConvertIsolationLevel(t *testing.T) {
 	if err == nil {
 		t.Fatal("must fail but it didn't")
 	}
+}
+
+// equalErrors is a helper function that compares two errors
+// by comparing their nilness, underlying type, and Error messages
+func equalErrors(e1 error, e2 error) bool {
+	if e1 == nil && e2 == nil {
+		return true
+	}
+	if e1 == nil && e2 != nil || e1 != nil && e2 == nil {
+		return false
+	}
+	return reflect.TypeOf(e1) == reflect.TypeOf(e2) &&
+		e1.Error() == e2.Error()
+}
+
+// TestCheckBadConn verifies that different combinations of
+// configuration, inputs and errors result in the proper output
+// error and connection state.
+func TestCheckBadConn(t *testing.T) {
+
+	netErr := &net.OpError{Err: fmt.Errorf("fake net.Error")}
+	streamErr := StreamError{Message: "fake StreamError"}
+	serverErr := ServerError{sqlError: Error{Message: "fake ServerError"}}
+	goodConnErr := fmt.Errorf("fake error that leaves connection good")
+
+	testInputs := []struct {
+		err              error
+		mayRetry         bool
+		disableRetry     bool
+		expectedErr      error
+		expectedConnGood bool
+	}{
+		{nil, false, false, nil, true},
+		{nil, true, false, nil, true},
+		{nil, false, true, nil, true},
+		{nil, true, true, nil, true},
+		{io.EOF, false, false, io.EOF, false},
+		{io.EOF, true, false, newRetryableError(io.EOF), false},
+		{io.EOF, false, true, io.EOF, false},
+		{io.EOF, true, true, io.EOF, false},
+		{netErr, false, false, netErr, false},
+		{netErr, true, false, newRetryableError(netErr), false},
+		{netErr, false, true, netErr, false},
+		{netErr, true, true, netErr, false},
+		{streamErr, false, false, streamErr, false},
+		{streamErr, true, false, newRetryableError(streamErr), false},
+		{streamErr, false, true, streamErr, false},
+		{streamErr, true, true, streamErr, false},
+		{serverErr, false, false, serverErr, false},
+		{serverErr, true, false, newRetryableError(serverErr), false},
+		{serverErr, false, true, serverErr, false},
+		{serverErr, true, true, serverErr, false},
+		{goodConnErr, false, false, goodConnErr, true},
+		{goodConnErr, true, false, goodConnErr, true},
+		{goodConnErr, false, true, goodConnErr, true},
+		{goodConnErr, true, true, goodConnErr, true},
+	}
+
+	c := Conn{connector: &Connector{params: msdsn.Config{}}}
+
+	for _, ti := range testInputs {
+		c.connectionGood = true
+		c.connector.params.DisableRetry = ti.disableRetry
+		actualErr := c.checkBadConn(ti.err, ti.mayRetry)
+		if !equalErrors(actualErr, ti.expectedErr) ||
+			c.connectionGood != ti.expectedConnGood {
+			t.Fatalf("checkBadConn returned unexpected result for input err = '%+v', mayRetry = '%t', disableRetry = '%t': "+
+				"Got output err = '%+v', connectionGood = '%t', "+
+				"wanted output err = '%+v', connectionGood = '%t'",
+				ti.err, ti.mayRetry, ti.disableRetry, actualErr, c.connectionGood, ti.expectedErr, ti.expectedConnGood)
+		}
+	}
+
+	// This must be the final test in this function, because we expect it to panic
+	defer func() { recover() }()
+	c.checkBadConn(driver.ErrBadConn, true)
+	t.Fatalf("checkBadConn did not panic as expected when passed driverErrBadConn")
 }

--- a/mssql_test.go
+++ b/mssql_test.go
@@ -142,7 +142,7 @@ func TestCheckBadConn(t *testing.T) {
 		if !equalErrors(actualErr, ti.expectedErr) ||
 			c.connectionGood != ti.expectedConnGood {
 			t.Fatalf("checkBadConn returned unexpected result for input err = '%+v', mayRetry = '%t', disableRetry = '%t': "+
-				"Got output err = '%+v', connectionGood = '%t', "+
+				"got output err = '%+v', connectionGood = '%t', "+
 				"wanted output err = '%+v', connectionGood = '%t'",
 				ti.err, ti.mayRetry, ti.disableRetry, actualErr, c.connectionGood, ti.expectedErr, ti.expectedConnGood)
 		}
@@ -152,4 +152,79 @@ func TestCheckBadConn(t *testing.T) {
 	defer func() { recover() }()
 	c.checkBadConn(driver.ErrBadConn, true)
 	t.Fatalf("checkBadConn did not panic as expected when passed driverErrBadConn")
+}
+
+// TestBadConnRejection verifies that database operations that start
+// with a bad connection fail with the sentinel error driver.ErrBadConn.
+// That instructs the database/sql connection pool logic to discard the
+// bad connection and, if appropriate, attempt to retry the operation
+// with another connection.
+func TestBadConnRejection(t *testing.T) {
+
+	c := Conn{connectionGood: false}
+
+	if _, err := c.Begin(); err != driver.ErrBadConn {
+		t.Fatalf("Begin did not fail on bad connection: "+
+			"got err = '%+v', wanted err = '%+v'", err, driver.ErrBadConn)
+	}
+
+	if err := c.Commit(); err != driver.ErrBadConn {
+		t.Fatalf("Commit did not fail on bad connection: "+
+			"got err = '%+v', wanted err = '%+v'", err, driver.ErrBadConn)
+	}
+
+	if err := c.Rollback(); err != driver.ErrBadConn {
+		t.Fatalf("Rollback did not fail on bad connection: "+
+			"got err = '%+v', wanted err = '%+v'", err, driver.ErrBadConn)
+	}
+
+	if _, err := c.Prepare("query"); err != driver.ErrBadConn {
+		t.Fatalf("Prepare did not fail on bad connection: "+
+			"got err = '%+v', wanted err = '%+v'", err, driver.ErrBadConn)
+	}
+
+	if _, err := c.PrepareContext(context.Background(), "query"); err != driver.ErrBadConn {
+		t.Fatalf("PrepareContext did not fail on bad connection: "+
+			"got err = '%+v', wanted err = '%+v'", err, driver.ErrBadConn)
+	}
+
+	if err := c.Ping(context.Background()); err != driver.ErrBadConn {
+		t.Fatalf("Ping did not fail on bad connection: "+
+			"got err = '%+v', wanted err = '%+v'", err, driver.ErrBadConn)
+	}
+
+	if _, err := c.BeginTx(context.Background(), driver.TxOptions{}); err != driver.ErrBadConn {
+		t.Fatalf("BeginTx did not fail on bad connection: "+
+			"got err = '%+v', wanted err = '%+v'", err, driver.ErrBadConn)
+	}
+
+	s := &Stmt{c: &c}
+
+	if _, err := s.Query(nil); err != driver.ErrBadConn {
+		t.Fatalf("Query did not fail on bad connection: "+
+			"got err = '%+v', wanted err = '%+v'", err, driver.ErrBadConn)
+	}
+
+	if _, err := s.QueryContext(context.Background(), nil); err != driver.ErrBadConn {
+		t.Fatalf("QueryContext did not fail on bad connection: "+
+			"got err = '%+v', wanted err = '%+v'", err, driver.ErrBadConn)
+	}
+
+	if _, err := s.Exec(nil); err != driver.ErrBadConn {
+		t.Fatalf("Exec did not fail on bad connection: "+
+			"got err = '%+v', wanted err = '%+v'", err, driver.ErrBadConn)
+	}
+
+	if _, err := s.ExecContext(context.Background(), nil); err != driver.ErrBadConn {
+		t.Fatalf("ExecContext did not fail on bad connection: "+
+			"got err = '%+v', wanted err = '%+v'", err, driver.ErrBadConn)
+	}
+
+	r := &Rows{stmt: s}
+
+	if err := r.Next(nil); err != driver.ErrBadConn {
+		t.Fatalf("Next did not fail on bad connection: "+
+			"got err = '%+v', wanted err = '%+v'", err, driver.ErrBadConn)
+	}
+
 }

--- a/queries_test.go
+++ b/queries_test.go
@@ -2277,26 +2277,38 @@ func TestDisconnect3(t *testing.T) {
 	checkConnStr(t)
 	SetLogger(testLogger{t})
 
-	// Revert to the normal dialer after the test is done.
+	// Revert to the normal dialer creation function after the test is done.
 	normalCreateDialer := createDialer
 	defer func() {
 		createDialer = normalCreateDialer
 	}()
 
-	// Create a dialer we can interrupt to simulate a broken connection
-	disrupt := make(chan bool)
-	waitDisrupt := make(chan bool)
+	// Implement a dialer creation function that returns interruptable dialers.
+	// These dialers make connections that we can interrupt to simulate breakage.
+	// The driver may call our dialer creation function multiple times to create
+	// multiple independent dialers, so we must signal all of them to ensure that
+	// the dialer controlling our test connection interrupts it.
+	var wg sync.WaitGroup
+	var mu sync.Mutex
+	makeInterruptDialers := true
+	disconnect := make(chan bool)
 	createDialer = func(p *msdsn.Config) Dialer {
 		nd := netDialer{&net.Dialer{Timeout: p.DialTimeout, KeepAlive: p.KeepAlive}}
-		di := &dialerInterrupt{nd: nd}
-		go func() {
-			// Wait for signal, then break channel in both directions
-			<-disrupt
-			di.Interrupt(true)
-			di.Interrupt(false)
-			waitDisrupt <- true
-		}()
-		return di
+		mu.Lock()
+		defer mu.Unlock()
+		if makeInterruptDialers {
+			di := &dialerInterrupt{nd: nd}
+			wg.Add(1)
+			go func() {
+				// Wait for signal, then break connection in both directions
+				<-disconnect
+				di.Interrupt(true)
+				di.Interrupt(false)
+				wg.Done()
+			}()
+			return di
+		}
+		return nd
 	}
 
 	// Disable retry to provide behavior requested in #275
@@ -2324,9 +2336,14 @@ func TestDisconnect3(t *testing.T) {
 			result, err, desiredResult, nil)
 	}
 
-	// Signal to break the connection and wait for it to be broken
-	disrupt <- true
-	<-waitDisrupt
+	// Stop making interuptable dialers, because we have already established
+	// our interruptable test connection, then signal and wait for our test
+	// connection to be broken
+	mu.Lock()
+	makeInterruptDialers = false
+	mu.Unlock()
+	close(disconnect)
+	wg.Wait()
 
 	// Broken connection should cause immediate and final failure
 	result = "<none>"
@@ -2352,26 +2369,38 @@ func TestDisconnect4(t *testing.T) {
 	checkConnStr(t)
 	SetLogger(testLogger{t})
 
-	// Revert to the normal dialer after the test is done.
+	// Revert to the normal dialer creation function after the test is done.
 	normalCreateDialer := createDialer
 	defer func() {
 		createDialer = normalCreateDialer
 	}()
 
-	// Create a dialer we can interrupt to simulate a broken connection
-	disrupt := make(chan bool)
-	waitDisrupt := make(chan bool)
+	// Implement a dialer creation function that returns interruptable dialers.
+	// These dialers make connections that we can interrupt to simulate breakage.
+	// The driver may call our dialer creation function multiple times to create
+	// multiple independent dialers, so we must signal all of them to ensure that
+	// the dialer controlling our test connection interrupts it.
+	var wg sync.WaitGroup
+	var mu sync.Mutex
+	makeInterruptDialers := true
+	disconnect := make(chan bool)
 	createDialer = func(p *msdsn.Config) Dialer {
 		nd := netDialer{&net.Dialer{Timeout: p.DialTimeout, KeepAlive: p.KeepAlive}}
-		di := &dialerInterrupt{nd: nd}
-		go func() {
-			// Wait for signal, then break channel in both directions
-			<-disrupt
-			di.Interrupt(true)
-			di.Interrupt(false)
-			waitDisrupt <- true
-		}()
-		return di
+		mu.Lock()
+		defer mu.Unlock()
+		if makeInterruptDialers {
+			di := &dialerInterrupt{nd: nd}
+			wg.Add(1)
+			go func() {
+				// Wait for signal, then break connection in both directions
+				<-disconnect
+				di.Interrupt(true)
+				di.Interrupt(false)
+				wg.Done()
+			}()
+			return di
+		}
+		return nd
 	}
 
 	// Enable retry to provide behavior requested in #586
@@ -2399,9 +2428,14 @@ func TestDisconnect4(t *testing.T) {
 			result, err, desiredResult, nil)
 	}
 
-	// Signal to break the connection and wait for it to be broken
-	disrupt <- true
-	<-waitDisrupt
+	// Stop making interuptable dialers, because we have already established
+	// our interruptable test connection, then signal and wait for our test
+	// connection to be broken
+	mu.Lock()
+	makeInterruptDialers = false
+	mu.Unlock()
+	close(disconnect)
+	wg.Wait()
 
 	// Broken connection should cause the next query to initially fail internally,
 	// but then the logic within database/sql should transparently discard the bad

--- a/queries_test.go
+++ b/queries_test.go
@@ -5,7 +5,6 @@ import (
 	"context"
 	"database/sql"
 	"database/sql/driver"
-	"errors"
 	"fmt"
 	"io"
 	"log"
@@ -2285,10 +2284,22 @@ func TestDisconnect3(t *testing.T) {
 	}()
 
 	// Create a dialer we can interrupt to simulate a broken connection
-	var di *dialerInterrupt
+	disrupt := make(chan bool)
+	defer close(disrupt)
+	waitDisrupt := make(chan bool)
+	defer close(waitDisrupt)
 	createDialer = func(p *msdsn.Config) Dialer {
 		nd := netDialer{&net.Dialer{Timeout: p.DialTimeout, KeepAlive: p.KeepAlive}}
-		di = &dialerInterrupt{nd: nd}
+		di := &dialerInterrupt{nd: nd}
+		go func() {
+			_, ok := <-disrupt
+			if ok {
+				// Break channel in both directions
+				di.Interrupt(true)
+				di.Interrupt(false)
+				waitDisrupt <- true
+			}
+		}()
 		return di
 	}
 
@@ -2317,9 +2328,9 @@ func TestDisconnect3(t *testing.T) {
 			result, err, desiredResult, nil)
 	}
 
-	// Simulate breaking the connection
-	di.Interrupt(true)
-	di.Interrupt(false)
+	// Signal to break the connection and wait for it to be broken
+	disrupt <- true
+	<-waitDisrupt
 
 	// Broken connection should cause immediate and final failure
 	result = "<none>"
@@ -2352,10 +2363,22 @@ func TestDisconnect4(t *testing.T) {
 	}()
 
 	// Create a dialer we can interrupt to simulate a broken connection
-	var di *dialerInterrupt
+	disrupt := make(chan bool)
+	defer close(disrupt)
+	waitDisrupt := make(chan bool)
+	defer close(waitDisrupt)
 	createDialer = func(p *msdsn.Config) Dialer {
 		nd := netDialer{&net.Dialer{Timeout: p.DialTimeout, KeepAlive: p.KeepAlive}}
-		di = &dialerInterrupt{nd: nd}
+		di := &dialerInterrupt{nd: nd}
+		go func() {
+			_, ok := <-disrupt
+			if ok {
+				// Break channel in both directions
+				di.Interrupt(true)
+				di.Interrupt(false)
+				waitDisrupt <- true
+			}
+		}()
 		return di
 	}
 
@@ -2384,9 +2407,9 @@ func TestDisconnect4(t *testing.T) {
 			result, err, desiredResult, nil)
 	}
 
-	// Break the connection
-	di.Interrupt(true)
-	di.Interrupt(false)
+	// Signal to break the connection and wait for it to be broken
+	disrupt <- true
+	<-waitDisrupt
 
 	// Broken connection should cause the next query to initially fail internally,
 	// but then the logic within database/sql should transparently discard the bad
@@ -2434,10 +2457,9 @@ func TestDisconnect5(t *testing.T) {
 
 	// Break the connection during a query by raising a fatal error in SQL Server
 	_, err = db.Exec("raiserror('fatal error', 20, 1) with log")
-	var sqlErr Error
-	if !errors.As(err, &sqlErr) || sqlErr.Number != 596 {
-		t.Fatalf("Failed to break connection. Got err = '%v', wanted error = '%v'",
-			err, "Cannot continue the execution because the session is in the kill state.")
+	if _, ok := err.(ServerError); !ok {
+		t.Fatalf("Failed to break connection. Got err = '%#v', wanted error = '%#v'",
+			err, ServerError{})
 	}
 
 	// The broken connection should have been removed from the pool and replaced

--- a/queries_test.go
+++ b/queries_test.go
@@ -2486,6 +2486,16 @@ func TestDisconnect5(t *testing.T) {
 	_, isServerError := err.(ServerError)
 	_, isNetError := err.(*net.OpError)
 	if !isServerError && !isNetError {
+		// There are a variety of errors that could be returned for a broken
+		// connection. The majority should fall within the mssql.ServerError
+		// and net.OpError categories. We try to make our test more specific
+		// by accepting only those errors, but that makes this test somewhat
+		// less resilient than desired. If this test fails because we start
+		// seeing previously unseen errors and those new errors also indicate
+		// a broken connection, then add them above. If this happens regularly,
+		// consider just allowing any error here. That will reduce any false
+		// alarms from this test at the cost of reduced ability to detect
+		// genuine test failure.
 		t.Fatalf("Failed to break connection. Got err = '%#v', wanted error = '%s'",
 			err, "mssql.ServerError or net.OpError")
 	}

--- a/queries_test.go
+++ b/queries_test.go
@@ -2295,10 +2295,11 @@ func TestDisconnect3(t *testing.T) {
 	// Disable retry to provide behavior requested in #275
 	cp := testConnParams(t)
 	cp.DisableRetry = true
+	connStr := cp.URL().String()
 
 	// Open a connection pool that holds a single connection to make sure all
 	// queries run on the same connection.
-	db, err := sql.Open("sqlserver", cp.URL().String())
+	db, err := sql.Open("sqlserver", connStr)
 	if err != nil {
 		t.Log(err)
 	}
@@ -2361,10 +2362,11 @@ func TestDisconnect4(t *testing.T) {
 	// Enable retry to provide behavior requested in #586
 	cp := testConnParams(t)
 	cp.DisableRetry = false
+	connStr := cp.URL().String()
 
 	// Open a connection pool that holds a single connection to make sure all
 	// queries run on the same connection.
-	db, err := sql.Open("sqlserver", makeConnStr(t).String())
+	db, err := sql.Open("sqlserver", connStr)
 	if err != nil {
 		t.Log(err)
 	}

--- a/queries_test.go
+++ b/queries_test.go
@@ -5,6 +5,7 @@ import (
 	"context"
 	"database/sql"
 	"database/sql/driver"
+	"errors"
 	"fmt"
 	"io"
 	"log"
@@ -2263,6 +2264,190 @@ func TestDisconnect2(t *testing.T) {
 		}
 	case <-timeout:
 		t.Fatal("timeout")
+	}
+}
+
+// TestDisconnect3 verifies the bad connection behavior defined in
+// response to issue #275. In this mode a query that starts on a bad
+// connection should always fail immediately, returning a detailed
+// error message.
+func TestDisconnect3(t *testing.T) {
+	if testing.Short() {
+		t.Skip("short")
+	}
+	checkConnStr(t)
+	SetLogger(testLogger{t})
+
+	// Revert to the normal dialer after the test is done.
+	normalCreateDialer := createDialer
+	defer func() {
+		createDialer = normalCreateDialer
+	}()
+
+	// Create a dialer we can interrupt to simulate a broken connection
+	var di *dialerInterrupt
+	createDialer = func(p *msdsn.Config) Dialer {
+		nd := netDialer{&net.Dialer{Timeout: p.DialTimeout, KeepAlive: p.KeepAlive}}
+		di = &dialerInterrupt{nd: nd}
+		return di
+	}
+
+	// Disable retry to provide behavior requested in #275
+	cp := testConnParams(t)
+	cp.DisableRetry = true
+
+	// Open a connection pool that holds a single connection to make sure all
+	// queries run on the same connection.
+	db, err := sql.Open("sqlserver", cp.URL().String())
+	if err != nil {
+		t.Log(err)
+	}
+	defer db.Close()
+	db.SetMaxIdleConns(1)
+	db.SetMaxOpenConns(1)
+
+	// Demonstrate connection is good the first time we use it
+	result := "<none>"
+	desiredResult := "before disconnect"
+	row := db.QueryRow(fmt.Sprintf("select '%s'", desiredResult))
+	if err = row.Scan(&result); err != nil || result != desiredResult {
+		t.Fatalf("Connection bad on first use. "+
+			"Got result = '%s', err = '%v', wanted result = '%s', err = '%v'",
+			result, err, desiredResult, nil)
+	}
+
+	// Simulate breaking the connection
+	di.Interrupt(true)
+	di.Interrupt(false)
+
+	// Broken connection should cause immediate and final failure
+	result = "<none>"
+	desiredErr := "failed to send SQL Batch: disconnect"
+	row = db.QueryRow("select 'after disconnect'")
+	if err = row.Scan(&result); err == nil || err.Error() != desiredErr {
+		t.Fatalf("Connection did not fail as expected. "+
+			"Got result = '%s', err = '%v', wanted result = '%s', err = '%v'",
+			result, err, "<none>", desiredErr)
+	}
+}
+
+// TestDisconnect4 verifies the bad connection behavior defined in
+// response to issue #586. In this mode a query that starts on a bad
+// connection should fail with an error that allows the automatic retry
+// logic within database/sql to retry the query with a new connection,
+// which then succeeds. From the perspective of the consuming application,
+// there is no error.
+func TestDisconnect4(t *testing.T) {
+	if testing.Short() {
+		t.Skip("short")
+	}
+	checkConnStr(t)
+	SetLogger(testLogger{t})
+
+	// Revert to the normal dialer after the test is done.
+	normalCreateDialer := createDialer
+	defer func() {
+		createDialer = normalCreateDialer
+	}()
+
+	// Create a dialer we can interrupt to simulate a broken connection
+	var di *dialerInterrupt
+	createDialer = func(p *msdsn.Config) Dialer {
+		nd := netDialer{&net.Dialer{Timeout: p.DialTimeout, KeepAlive: p.KeepAlive}}
+		di = &dialerInterrupt{nd: nd}
+		return di
+	}
+
+	// Enable retry to provide behavior requested in #586
+	cp := testConnParams(t)
+	cp.DisableRetry = false
+
+	// Open a connection pool that holds a single connection to make sure all
+	// queries run on the same connection.
+	db, err := sql.Open("sqlserver", makeConnStr(t).String())
+	if err != nil {
+		t.Log(err)
+	}
+	defer db.Close()
+	db.SetMaxIdleConns(1)
+	db.SetMaxOpenConns(1)
+
+	// Demonstrate connection is good the first time we use it
+	result := "<none>"
+	desiredResult := "before disconnect"
+	row := db.QueryRow(fmt.Sprintf("select '%s'", desiredResult))
+	if err = row.Scan(&result); err != nil || result != desiredResult {
+		t.Fatalf("Connection bad on first use. "+
+			"Got result = '%s', err = '%v', wanted result = '%s', err = '%v'",
+			result, err, desiredResult, nil)
+	}
+
+	// Break the connection
+	di.Interrupt(true)
+	di.Interrupt(false)
+
+	// Broken connection should cause the next query to initially fail internally,
+	// but then the logic within database/sql should transparently discard the bad
+	// connection, retry the query with a new connection, and ultimately succeed.
+	result = "<none>"
+	desiredResult = "after disconnect"
+	row = db.QueryRow(fmt.Sprintf("select '%s'", desiredResult))
+	if err = row.Scan(&result); err != nil || result != desiredResult {
+		t.Fatalf("Connection bad on second use. "+
+			"Got result = '%s', err = '%v', wanted result = '%s', err = '%v'",
+			result, err, desiredResult, nil)
+	}
+}
+
+// TestDisconnect5 verifies that a connection that goes bad during a query results in:
+// 1. The query failing with a detailed error message
+// 2. The bad connection being immediately removed from the pool
+// 3. The subsequent query using a new connection and succeeding
+func TestDisconnect5(t *testing.T) {
+	if testing.Short() {
+		t.Skip("short")
+	}
+	checkConnStr(t)
+	SetLogger(testLogger{t})
+
+	// Open a connection pool that holds a single connection to make sure all
+	// queries run on the same connection.
+	db, err := sql.Open("sqlserver", makeConnStr(t).String())
+	if err != nil {
+		t.Log(err)
+	}
+	defer db.Close()
+	db.SetMaxIdleConns(1)
+	db.SetMaxOpenConns(1)
+
+	// Demonstrate connection is good the first time we use it
+	result := "<none>"
+	desiredResult := "before disconnect"
+	row := db.QueryRow(fmt.Sprintf("select '%s'", desiredResult))
+	if err = row.Scan(&result); err != nil || result != desiredResult {
+		t.Fatalf("Connection bad on first use. "+
+			"Got result = '%s', err = '%v', wanted result = '%s', err = '%v'",
+			result, err, desiredResult, nil)
+	}
+
+	// Break the connection during a query by raising a fatal error in SQL Server
+	_, err = db.Exec("raiserror('fatal error', 20, 1) with log")
+	var sqlErr Error
+	if !errors.As(err, &sqlErr) || sqlErr.Number != 596 {
+		t.Fatalf("Failed to break connection. Got err = '%v', wanted error = '%v'",
+			err, "Cannot continue the execution because the session is in the kill state.")
+	}
+
+	// The broken connection should have been removed from the pool and replaced
+	// with a new connection at the end of the previous query, allowing this next
+	// query to succeed.
+	result = "<none>"
+	desiredResult = "after disconnect"
+	row = db.QueryRow(fmt.Sprintf("select '%s'", desiredResult))
+	if err = row.Scan(&result); err != nil || result != desiredResult {
+		t.Fatalf("Connection bad on second use. "+
+			"Got result = '%s', err = '%v', wanted result = '%s', err = '%v'",
+			result, err, desiredResult, nil)
 	}
 }
 

--- a/queries_test.go
+++ b/queries_test.go
@@ -2285,20 +2285,16 @@ func TestDisconnect3(t *testing.T) {
 
 	// Create a dialer we can interrupt to simulate a broken connection
 	disrupt := make(chan bool)
-	defer close(disrupt)
 	waitDisrupt := make(chan bool)
-	defer close(waitDisrupt)
 	createDialer = func(p *msdsn.Config) Dialer {
 		nd := netDialer{&net.Dialer{Timeout: p.DialTimeout, KeepAlive: p.KeepAlive}}
 		di := &dialerInterrupt{nd: nd}
 		go func() {
-			_, ok := <-disrupt
-			if ok {
-				// Break channel in both directions
-				di.Interrupt(true)
-				di.Interrupt(false)
-				waitDisrupt <- true
-			}
+			// Wait for signal, then break channel in both directions
+			<-disrupt
+			di.Interrupt(true)
+			di.Interrupt(false)
+			waitDisrupt <- true
 		}()
 		return di
 	}
@@ -2364,20 +2360,16 @@ func TestDisconnect4(t *testing.T) {
 
 	// Create a dialer we can interrupt to simulate a broken connection
 	disrupt := make(chan bool)
-	defer close(disrupt)
 	waitDisrupt := make(chan bool)
-	defer close(waitDisrupt)
 	createDialer = func(p *msdsn.Config) Dialer {
 		nd := netDialer{&net.Dialer{Timeout: p.DialTimeout, KeepAlive: p.KeepAlive}}
 		di := &dialerInterrupt{nd: nd}
 		go func() {
-			_, ok := <-disrupt
-			if ok {
-				// Break channel in both directions
-				di.Interrupt(true)
-				di.Interrupt(false)
-				waitDisrupt <- true
-			}
+			// Wait for signal, then break channel in both directions
+			<-disrupt
+			di.Interrupt(true)
+			di.Interrupt(false)
+			waitDisrupt <- true
 		}()
 		return di
 	}
@@ -2457,9 +2449,11 @@ func TestDisconnect5(t *testing.T) {
 
 	// Break the connection during a query by raising a fatal error in SQL Server
 	_, err = db.Exec("raiserror('fatal error', 20, 1) with log")
-	if _, ok := err.(ServerError); !ok {
-		t.Fatalf("Failed to break connection. Got err = '%#v', wanted error = '%#v'",
-			err, ServerError{})
+	_, isServerError := err.(ServerError)
+	_, isNetError := err.(*net.OpError)
+	if !isServerError && !isNetError {
+		t.Fatalf("Failed to break connection. Got err = '%#v', wanted error = '%s'",
+			err, "mssql.ServerError or net.OpError")
 	}
 
 	// The broken connection should have been removed from the pool and replaced

--- a/queries_test.go
+++ b/queries_test.go
@@ -2332,7 +2332,7 @@ func TestDisconnect3(t *testing.T) {
 	row := db.QueryRow(fmt.Sprintf("select '%s'", desiredResult))
 	if err = row.Scan(&result); err != nil || result != desiredResult {
 		t.Fatalf("Connection bad on first use. "+
-			"Got result = '%s', err = '%v', wanted result = '%s', err = '%v'",
+			"Got result = '%s', err = '%v',  result = '%s', err = '%v'",
 			result, err, desiredResult, nil)
 	}
 

--- a/token.go
+++ b/token.go
@@ -75,6 +75,15 @@ const (
 	fedAuthInfoSPN    = 0x02
 )
 
+// serverError is returned when we receive a DONE token with DONE_SRVERROR flag.
+// It means an error was severe enough to require the result set to be discarded.
+// It was observed when raising fatal errors, with severity 20 or above.
+// These errors also close the connection.
+//
+// https://docs.microsoft.com/en-us/openspecs/sql_server_protocols/ms-sstds/bd77f9d0-6929-4fe1-b163-a525cd7cabd4
+// https://docs.microsoft.com/en-us/openspecs/sql_server_protocols/ms-sstds/90fb176f-83c6-4c9c-bc04-aaa8a400a5a1
+var serverError = errors.New("SQL Server had internal error")
+
 // COLMETADATA flags
 // https://msdn.microsoft.com/en-us/library/dd357363.aspx
 const (
@@ -699,7 +708,7 @@ func processSingleResponse(sess *tdsSession, ch chan tokenStruct, outs map[strin
 				sess.log.Printf("got DONE or DONEPROC status=%d", done.Status)
 			}
 			if done.Status&doneSrvError != 0 {
-				ch <- errors.New("SQL Server had internal error")
+				ch <- serverError
 				return
 			}
 			if sess.logFlags&logRows != 0 && done.Status&doneCount != 0 {

--- a/token.go
+++ b/token.go
@@ -82,7 +82,17 @@ const (
 //
 // https://docs.microsoft.com/en-us/openspecs/sql_server_protocols/ms-sstds/bd77f9d0-6929-4fe1-b163-a525cd7cabd4
 // https://docs.microsoft.com/en-us/openspecs/sql_server_protocols/ms-sstds/90fb176f-83c6-4c9c-bc04-aaa8a400a5a1
-var serverError = errors.New("SQL Server had internal error")
+type serverError struct {
+	sqlError Error
+}
+
+func (e serverError) Error() string {
+	return "SQL Server had internal error"
+}
+
+func (e serverError) Unwrap() error {
+	return e.sqlError
+}
 
 // COLMETADATA flags
 // https://msdn.microsoft.com/en-us/library/dd357363.aspx
@@ -708,7 +718,7 @@ func processSingleResponse(sess *tdsSession, ch chan tokenStruct, outs map[strin
 				sess.log.Printf("got DONE or DONEPROC status=%d", done.Status)
 			}
 			if done.Status&doneSrvError != 0 {
-				ch <- serverError
+				ch <- serverError{done.getError()}
 				return
 			}
 			if sess.logFlags&logRows != 0 && done.Status&doneCount != 0 {

--- a/token.go
+++ b/token.go
@@ -75,25 +75,6 @@ const (
 	fedAuthInfoSPN    = 0x02
 )
 
-// serverError is returned when we receive a DONE token with DONE_SRVERROR flag.
-// It means an error was severe enough to require the result set to be discarded.
-// It was observed when raising fatal errors, with severity 20 or above.
-// These errors also close the connection.
-//
-// https://docs.microsoft.com/en-us/openspecs/sql_server_protocols/ms-sstds/bd77f9d0-6929-4fe1-b163-a525cd7cabd4
-// https://docs.microsoft.com/en-us/openspecs/sql_server_protocols/ms-sstds/90fb176f-83c6-4c9c-bc04-aaa8a400a5a1
-type serverError struct {
-	sqlError Error
-}
-
-func (e serverError) Error() string {
-	return "SQL Server had internal error"
-}
-
-func (e serverError) Unwrap() error {
-	return e.sqlError
-}
-
 // COLMETADATA flags
 // https://msdn.microsoft.com/en-us/library/dd357363.aspx
 const (
@@ -718,7 +699,7 @@ func processSingleResponse(sess *tdsSession, ch chan tokenStruct, outs map[strin
 				sess.log.Printf("got DONE or DONEPROC status=%d", done.Status)
 			}
 			if done.Status&doneSrvError != 0 {
-				ch <- serverError{done.getError()}
+				ch <- ServerError{done.getError()}
 				return
 			}
 			if sess.logFlags&logRows != 0 && done.Status&doneCount != 0 {

